File: Tests/SignalR.Extras.Autofac.Test/LifetimeHubManagerFacts.cs
using System;
using System.Linq;

using Autofac;

using Xunit;

using static SignalR.Extras.Autofac.Test.TestHelper;


namespace SignalR.Extras.Autofac.Test
{

	public class LifetimeHubManagerFacts
	{

		[Fact]
		public void OnRegistrationThrowsIfContainerBuilderIsNull()
		{
			ContainerBuilder builder = null;
			Assert.Throws<ArgumentNullException>(() => builder.RegisterLifetimeHubManager());
		}


		[Fact]
		public void AutofacCanResolveLifetimeHubManager()
		{
			var container = SetupAutofacContainer();

			var lifetimeHubManager = container.Resolve<LifetimeHubManager>();
			Assert.NotNull(lifetimeHubManager);
		}


		[Fact]
		public void OnDisposingDisposesOwnedHubs()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			int disposedHubsCount = 0;
            var hubs = hubManager.ResolveHubs().OfType<ILifetimeHub>();
			foreach (var hub in hubs) {
				hub.OnDisposing += (o, a) => disposedHubsCount++;
			}
			var lifetimeHubManager = container.Resolve<LifetimeHubManager>();
			lifetimeHubManager.Dispose();
			Assert.Equal(2, disposedHubsCount);
		}

	}
}

File: Tests/SignalR.Extras.Autofac.Test/LifetimeHubManagerFacts.fs
namespace SignalR.Extras.Autofac.Test

module LifetimeHubManagerFacts =
    open System
    open System.Linq
    open Autofac
    open SignalR.Extras.Autofac
    open Xunit


    [<Fact>]
    let ``On Registration Throws If Container Builder Is Null``() =
        let builder : ContainerBuilder = null
        Assert.Throws<ArgumentNullException>(fun () -> builder.RegisterLifetimeHubManager())


    [<Fact>]
    let ``Autofac Can Resolve LifetimeHubManager``() =
        let container = TestHelper.SetupAutofacContainer()
        let lifetimeHubManager = container.Resolve<LifetimeHubManager>()
        Assert.NotNull(lifetimeHubManager)


    [<Fact>]
    let ``OnDisposing Disposes Owned Hubs``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        let disposedHubsCount = ref 0
        let hubs = hubManager.ResolveHubs().OfType<ILifetimeHub>()
        for hub in hubs do
            hub.OnDisposing.Add(fun(x) -> disposedHubsCount := !disposedHubsCount + 1)
        let lifetimeHubManager = container.Resolve<LifetimeHubManager>()
        lifetimeHubManager.Dispose()
        Assert.Equal(2, !disposedHubsCount)
